docs/proposals/kubermatic-installer.md

```diff
@@ -0,0 +1,128 @@
+  binary.
+* Use Helm 3 for managing releases.
+* Do not require the admin to manually use Helm to manage their installation.
+* Allow the admin to manually use Helm if they so choose. Do not hide things from an admin,
+  but make it clear that they are leaving the supported paths.
+* Provide clear, helpful log output and error messages.
+* Never perform dangerous (as in: can possibly lose data or change LoadBalancer IPs)
+  migration steps without warning the user first (e.g. "run again with --yes-migrate-x
+  to perform the migration").
+* Pin the installer to a specific Kubermatic version, i.e. to choose a different version
+  a different installer release must be downloaded.
+* Use the installer during E2E tests.
+
+### Non-Goals
+
+* Downgrades.
+* Interactive CLI/ncurses interface.
+* Migrate manual installations to be installer-managed.
+* Migrate Helm-based installations to the Kubermatic Operator.
+
+### Implementation
+
+The installation generally done sequentially. First experiments with doing it in parallel
+tasks yielded no good results:
+
+* It was difficult to print an understandable log for the user. Keeping the user informed
+  about what is happening to their cluster is important.
+* Error handling was difficult, especially in longer running parallel Helm invocations
+  and dependent things.
+* It makes reasoning when writing migration steps much harder if you have to very
+  carefully plan your migration code in all the goroutines that run during an upgrade.
+
+Having sequential steps made the code much easier to read and maintain. It sacrifices some
+speed for ease of use.
+
+During `./installer deploy`, the chosen "stack" is installed. By default (and currently
+because it's the only one) this is the "kubermatic" stack. The stack then does whatever
+it needs to run. For Kubermatic this means
+
+* setting up a "kubermatic-fast" StorageClass; this is a leftover from very early installer
+  versions and requires us to know the cloud provider. This requires yet another setting
+  to be made from the user, so maybe we should instead just inform the user that they need
+  to create a StorageClass?
+* installing cert-manager, nginx-ingress-controller, Dex and the Kubermatic Operator
+  Helm charts. "Installing" can also mean an upgrade. If a failed release is encountered,
+  the release is purged first, otherwise another install with Helm is impossible. If the
+  release in the cluster is identical to the Helm chart that is shipped with the installer,
+  nothing happens unless the user runs the installer with `--force`.
+* `kubectl apply` the given `KubermaticConfiguration`.
+
+If anything goes wrong during this, the installation is aborted. The user then can fix
+the problem and re-run the installer.
+
+It would be possible to also manage Seeds while we're at it, but the win between having to
+explain to the user that the installer can also install the seeds, vs. just making them
+`kubectl apply` them is questionable.
```

Comment:
Fair point. The advantage I would see though is to have an all in one type of installation and not having a "mix". Would an optional flag make sense here?

Reply:
I think this is a bridge we can burn when we come to it. This does not affect the rest of the installer in any way. :-)